### PR TITLE
*: support bootstrap cluster.

### PR DESCRIPTION
**Pull request by siddontang**
@ngaut @disksing

### CODE REVIEW

File: server/cluster.go [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+		nodes:       make(map[uint64]protopb.Node),
+		stores:      make(map[uint64]protopb.Store),
+	}
+
+	return c
+}
+
+func (s *Server) getClusterRootPath(clusterID uint64) string {
+	return strings.Join(
+		[]string{s.cfg.RootPath, "raft", string(uint64ToBytes(clusterID))},
```
Comment (disksing) [inline server/cluster.go]: Maybe you want to use `atoi` instead of `uint64ToBytes`. It may produce invalid chars such as `/`.

Comment (siddontang) [inline server/cluster.go]: no, we don't care it, it is safe here even it contains `/`.

File: server/cluster.go [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+	}
+
+	log.Infof("bootstrap cluster %d ok", clusterID)
+
+	s.clusterLock.Lock()
+	defer s.clusterLock.Unlock()
+
+	if _, ok := s.clusters[clusterID]; ok {
+		// We have bootstrapped cluster ok, and another goroutine quickly requests use this
+		// so we create the cluster object.
```
Comment (disksing) [inline server/cluster.go]: In that case, it seems the etcd transaction will not success.

File: server/cluster_test.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+	go s.svr.Run()
+}
+
+func (s *testClusterSuite) TearDownSuite(c *C) {
+	s.svr.Close()
+	s.client.Close()
+}
+
+func (s *testClusterSuite) allocID(c *C) uint64 {
+	id, err := s.svr.idAlloc.Alloc()
```
Comment (disksing) [inline server/cluster_test.go]: Should it alloc id by sending request to server?

Comment (siddontang) [inline server/cluster_test.go]: no need, allocator can do it directly.

File: server/cluster.go (hunk 2 above)
Comment (siddontang) [inline server/cluster.go]: no, the etcd transaction successes already. 

When we bootstrap in etcd, another thread can use it quickly and if it finds no cluster object is created, it creates. 

But It is nearly impossible to enter this case, I will log it later.

**Comment by disksing** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @ngaut

File: server/cluster.go [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+// Operation list, pd can only handle operations like auto-balance, split,
+// merge sequentially, and every operation will be assigned a unique incremental ID.
+// pending list -> /raft/1/j/1, /raft/1/j/2, value is operation job.
+//
+// Encode region search key:
+//  1, the maximum end region key is empty, so the encode key is \xFF
+//  2, other region end key is not empty, the encode key is \z end_key
+
+type raftCluster struct {
+	sync.Mutex
```
Comment (ngaut) [inline server/cluster.go]: which field does Mutex protect ? Please add comments.

Comment (siddontang) [inline server/cluster.go]: For later use, I will remove it now.

File: server/cluster.go (hunk 1 above)
Comment (ngaut) [inline server/cluster.go]: I prefer atoi too.

**Comment by siddontang** [comment]
PTAL @ngaut

File: server/cluster.go [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+func encodeRegionSearchKey(endKey []byte) string {
+	if len(endKey) == 0 {
+		return "\xFF"
+	}
+
+	return string(append([]byte{'z'}, endKey...))
+}
+
+func getClusterNodeKey(clusterRootPath string, nodeID uint64) string {
```
Comment (ngaut) [inline server/cluster.go]: s/getClusterNodeKey/makeNodeKey/g

File: server/cluster.go [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+	return strings.Join([]string{clusterRootPath, "n", strconv.FormatUint(nodeID, 10)}, "/")
+}
+
+func getClusterStoreKey(clusterRootPath string, storeID uint64) string {
```
Comment (ngaut) [inline server/cluster.go]: Ditto.

File: server/cluster.go [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+	return strings.Join([]string{clusterRootPath, "s", strconv.FormatUint(storeID, 10)}, "/")
+}
+
+func getClusterRegionKey(clusterRootPath string, regionID uint64) string {
```
Comment (ngaut) [inline server/cluster.go]: Ditto.

File: server/cluster.go [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+	return strings.Join([]string{clusterRootPath, "r", strconv.FormatUint(regionID, 10)}, "/")
+}
+
+func getClusterRegionSearchKey(clusterRootPath string, endKey []byte) string {
```
Comment (ngaut) [inline server/cluster.go]: Ditto.

File: server/cluster.go [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+func (s *Server) bootstrapCluster(clusterID uint64, req *protopb.BootstrapRequest) error {
+	log.Infof("try to bootstrap cluster %d with %v", clusterID, req)
+
+	if err := checkBootstrapRequest(clusterID, req); err != nil {
+		return errors.Trace(err)
+	}
+
+	var ops []clientv3.Op
```
Comment (ngaut) [inline server/cluster.go]: Move down.

Comment (siddontang) [inline server/cluster.go]: move to which line?

Comment (ngaut) [inline server/cluster.go]: var ops []clientv3.Op

**Comment by ngaut** [comment]
LGTM

**Pull request merged**